### PR TITLE
23.07 Release candidate

### PR DESCRIPTION
**Pull request by odudex**
This is draft pull request made to highlight the candidate code for the next official release.
Several features, tools and optimizations were created, from battery indicator to encrypted storage of mnemonics, they will be listed in more details later.
The goal to be met for a release is to increase pytest code coverage to 95% and update documentation to cover new and modified  features.
If you would like to help test this development branch you can build it from source, or flash binaries available [here](https://github.com/odudex/krux_binaries)

### CODE REVIEW

File: docs/getting-started/generating-a-mnemonic.en.md
Comment (jdlcdl) [inline docs/getting-started/generating-a-mnemonic.en.md:54]: Would "..., which contain each pixel in rgb565 format, ..." (or similar) be more informative and still read well?

File: docs/getting-started/generating-a-mnemonic.en.md
Comment (jdlcdl) [inline docs/getting-started/generating-a-mnemonic.en.md:1]: "to outsources entropy" becomes "to outsource entropy"
and
"camera as source of" becomes "camera as a source of"

**Comment by tadeubas** [comment]
Hi @jdlcdl thx for pointing out these changes, plz create a PR to facilitate the incorporation of those!

**Comment by jdlcdl** [comment]
> Hi @jdlcdl thx for pointing out these changes, plz create a PR to facilitate the incorporation of those!

I might not get to doing that soon, so in the meantime, I'm just doing my best to follow latest changes and was hoping that leaving notes could help the authors throw improvements that catch my eye into their next commit.  Thanks for the nudge, I'll be getting to that as soon as I have time.

**Comment by odudex** [comment]
Thanks Jean! I'll fix it. Any type of help and review is really welcome!

File: docs/getting-started/tools.en.md
Comment (jreesun) [inline docs/getting-started/tools.en.md:18]: If I had an SD card plugged in, but my mnemonic was stored on the flash, would this try to wipe both places or just the SD card?

Ideally it would always try to wipe both, I think.

File: pyproject.toml
Comment (jreesun) [inline pyproject.toml:25]: needs to be updated to 23.08 (or 23.09 if that ends up happening, ha)

File: src/boot.py
Comment (jreesun) [inline src/boot.py:94]: a constant here would be better

File: src/boot.py
Comment (jreesun) [inline src/boot.py:126]: `+ MIN_SPLASH_TIME`

File: src/krux/camera.py
Comment (jreesun) [inline src/krux/camera.py:167]: Yeah, this is too naive to use imo. Would be better to build entropy by having the user take a series of captures like they do a series of dice rolls. Perhaps even a MIN_CAPTURES analogous to MIN_ROLLS, if we assume each image is 10 bits of entropy...

Though ideally we'd be trying to count the entropy in each captured image.

File: src/krux/encryption.py
Comment (jreesun) [inline src/krux/encryption.py:50]: This makes me nervous since I'm not an expert and can't really audit this. Also because we're relying on the ucryptolib which I believe was implemented by MaixPy employees...

Encryption is easy to get wrong if you don't know what you're doing. How confident are you in this implementation? Did another set of eyes look this over?

File: src/krux/metadata.py
Comment (jreesun) [inline src/krux/metadata.py:22]: same here

File: src/krux/pages/files_manager.py
Comment (jreesun) [inline src/krux/pages/files_manager.py:1]: I would just call this `filemanager.py` or `file_manager.py` so it matches the class name like other files

File: src/krux/pages/files_manager.py
Comment (jreesun) [inline src/krux/pages/files_manager.py:38]: wrong description?

File: src/krux/pages/files_manager.py
Comment (jreesun) [inline src/krux/pages/files_manager.py:149]: dead code

File: src/krux/printers/cnc.py
Comment (jreesun) [inline src/krux/printers/cnc.py:264]: Maybe add a TODO: implement here, or make a github issue saying it needs to be implemented

File: src/krux/psbt.py
Comment (jreesun) [inline src/krux/psbt.py:49]: dead code

**Review by jreesun — CHANGES_REQUESTED** [review]
I might try to make another pass at this again if I get time since there's a lot to get through.

Hope you find my comments helpful.

This is a monumental change, and you're doing great!

File: CHANGELOG.md
Comment (odudex) [reply CHANGELOG.md:20]: I believe it was always 1 to 2048, I don't remember changing this. It's also how it is defined in the BIP39:
https://github.com/bitcoin/bips/blob/master/bip-0039/english.txt

File: docs/getting-started/tools.en.md
Comment (tadeubas) [reply docs/getting-started/tools.en.md:18]: This tool list all encrypted keys and allow you to wipe them one by one, independent of where it is stored (flash or SD card)

File: src/krux/camera.py
Comment (odudex) [reply src/krux/camera.py:167]: Makes no sense for me. Each pixel has 16 bits of information, its entropy may be low quality, non linear, with possible saturated regions, etc., but there are 76.800 of them, in a raw, uncompressed image.

File: src/krux/encryption.py
Comment (odudex) [reply src/krux/encryption.py:50]: I'm not an expert either, but Maixpy implementation seems reliable and following AES standards.
It's also working on Android App, which uses [Pycryptodome](https://pypi.org/project/pycryptodome/) and encryption from both are interoperable.
I posted an encrypted mnemonic QR code online with some sats to check if someone could hack it, no one did. A month ago a Brazilian youtube channel of education Bitcoin content with thousands of followers did the same challenge using Krux, no one hacked it too. Code and basic information were also provided.
The encryption methods are known and widely used, but off course it would be excellent to have more scrutiny on how it was implemented.

File: CHANGELOG.md
Comment (tadeubas) [reply CHANGELOG.md:20]: As I said on another comment: Available backup schemes (paper or metal) are focused on common users, they always start from the BIP39 word number (1-2048).

File: docs/faq.en.md
Comment (tadeubas) [reply docs/faq.en.md:62]: I've done that it is on the PR now

File: codecov.yml
Comment (tadeubas) [reply codecov.yml:1]: I've done this, on PR now

File: codecov.yml
Comment (tadeubas) [reply codecov.yml:5]: I've done this, on PR now

File: CHANGELOG.md
Comment (tadeubas) [reply CHANGELOG.md:2]: I've done this, on PR now

File: CHANGELOG.md
Comment (tadeubas) [reply CHANGELOG.md:13]: I've changed the text to say mnemonic words NUMBER, on PR now

File: CHANGELOG.md
Comment (tadeubas) [reply CHANGELOG.md:20]: I've changed the text to say mnemonic words NUMBER, on PR now

File: README.md
Comment (tadeubas) [reply README.md:293]: I've done this, on PR now

File: docs/faq.en.md
Comment (tadeubas) [reply docs/faq.en.md:51]: I've done this, on PR now

File: docs/getting-started/index.en.md
Comment (tadeubas) [reply docs/getting-started/index.en.md:1]: I've done this, on PR now

File: docs/getting-started/installing/from-gui.en.md
Comment (tadeubas) [reply docs/getting-started/installing/from-gui.en.md:1]: I've done this, on PR now

File: docs/getting-started/installing/from-test-release.en.md
Comment (tadeubas) [reply docs/getting-started/installing/from-test-release.en.md:1]: I've done this, on PR now

File: pyproject.toml
Comment (tadeubas) [reply pyproject.toml:25]: I've done this, on PR now

File: src/krux/metadata.py
Comment (tadeubas) [reply src/krux/metadata.py:22]: I've done this, on PR now

File: src/krux/pages/files_manager.py
Comment (tadeubas) [reply src/krux/pages/files_manager.py:38]: I've done this, on PR now

File: src/boot.py
Comment (tadeubas) [reply src/boot.py:94]: I've done this, on PR now

File: docs/getting-started/generating-a-mnemonic.en.md
Comment (odudex) [reply docs/getting-started/generating-a-mnemonic.en.md:54]: Something like [this](https://photos.app.goo.gl/29j64WVVbeBR7D578) could be done. But I can't affirm pixel colors and lightness variance are correlated to entropy's quality.
Here I was measuring 3 standard deviations (square root of variances):
l_stdev = lightness standard deviation
a_stdev = green–magenta standard deviation
b_stdev = blue–yellow standard deviation

Comment (jreesun) [reply docs/getting-started/generating-a-mnemonic.en.md:54]: So as not to hold up this PR any longer, could you mark the feature experimental until someone with experience in this domain can confirm that a single image will always provide sufficient entropy? My hunch is it won't, at least not all of the time, and I think Krux should have a way to indicate that to the user.

Entropy doesn't necessarily equal information. I can "generate" a 256 bit number from my mind, but it won't have 256 bits of entropy since my internal RNG isn't very good. I think pictures are similar, just because they have many bits of information doesn't mean that those bits are entropy.

**Review by jreesun — APPROVED** [review]

**Comment by odudex** [comment]
@jreesun ,sorry to do it after your approval, but there was some important recent changes @tadeubas made, from which I highlight a bugfix 75485e3bfc5b0e5f2b9655057c63724ade166c28.
I believe we can close now and this is the pack for next release.
I'm also working on other changes, in a [new branch on my repo](https://github.com/odudex/krux/tree/irq_based): some deep ones as to change the way buttons and touch works, to be event based, so clicks aren't lost when in high processing (capturing video or showing animated QR codes). But as this changes still require tests and simulator to be adapted let's leave them for next release.

File: docs/getting-started/generating-a-mnemonic.en.md
Comment (odudex) [reply docs/getting-started/generating-a-mnemonic.en.md:54]: Sorry just saw this now. I'll add a "experimental" tag on intro message.
Sorry I broke the link of experiments I made to estimate poor entropy. You can [try this post](https://twitter.com/odudex/status/1699126162650275945?s=20), from which we got good feedback. But this "detection" other improvements are things to be discussed for later releases.

**Pull request merged**
